### PR TITLE
Bug Fix: Segment Purger cannot purge old segments after schema evolution

### PR DESCRIPTION
**Pull request by sajjad-moradi**
Solves the issue https://github.com/apache/pinot/issues/10868

I updated the minion purge integration tests to add a scenario in which the segments are built with old schema. Without the fix that unit test fails.

### CODE REVIEW

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/PurgeMinionClusterIntegrationTest.java [hunk 1]
```diff
@@ -96,23 +98,27 @@ public void setUp()
...
     addTableConfig(purgeTableConfig);
     addTableConfig(purgeDeltaPassedTableConfig);
     addTableConfig(purgeDeltaNotPassedTableConfig);
+    addTableConfig(purgeOldSegmentsWithNewIndicesTableConfig);
 
     // Unpack the Avro files
     List<File> avroFiles = unpackAvroData(_tempDir);
 
-    // Create and upload segments
-    ClusterIntegrationTestUtils.buildSegmentsFromAvro(avroFiles, purgeTableConfig, schema, 0, _segmentDir1, _tarDir1);
```
Comment (PrachiKhobragade) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/PurgeMinionClusterIntegrationTest.java:107]: Why did we consolidate everything to same directory?

**Review by PrachiKhobragade — APPROVED** [review]

**Comment by Jackie-Jiang** [comment]
This doesn't seem to be the correct fix.

The real problem is that we are using the table config from the ZK and schema from the segment file (we picked this way before because schema might not be stored in ZK when this task was implemented). Instead, we should also read the schema from the ZK so that schema is in-sync with the table config. You may read the schema in `PurgeTaskExecutor` and pass it into the `SegmentPurger`, then use it to create the `SegmentGeneratorConfig`.

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/PurgeMinionClusterIntegrationTest.java (hunk 1 above)
Comment (sajjad-moradi) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/PurgeMinionClusterIntegrationTest.java:107]: That wasn't needed. The schema and table configs for all those three tables were the same, so were the segments that were built. So there's no need to build the segments three times.

**Comment by mcvsubbu** [comment]
> This doesn't seem to be the correct fix.
> 
> The real problem is that we are using the table config from the ZK and schema from the segment file (we picked this way before because schema might not be stored in ZK when this task was implemented). Instead, we should also read the schema from the ZK so that schema is in-sync with the table config. You may read the schema in `PurgeTaskExecutor` and pass it into the `SegmentPurger`, then use it to create the `SegmentGeneratorConfig`.

Minion purge task can purge data from old segments, which may have been built with an older schema (minus some columns). I thought the idea was to keep the old segments without the new column (and let the server auto-load default value into new column instead of baking it into the segment).

**Comment by sajjad-moradi** [comment]
> > This doesn't seem to be the correct fix.
> > The real problem is that we are using the table config from the ZK and schema from the segment file (we picked this way before because schema might not be stored in ZK when this task was implemented). Instead, we should also read the schema from the ZK so that schema is in-sync with the table config. You may read the schema in `PurgeTaskExecutor` and pass it into the `SegmentPurger`, then use it to create the `SegmentGeneratorConfig`.
> 
> Minion purge task can purge data from old segments, which may have been built with an older schema (minus some columns). I thought the idea was to keep the old segments without the new column (and let the server auto-load default value into new column instead of baking it into the segment).

Subbu, with fetching schema from ZK, once the segments are built by Minion and pushed to deep store, the Servers don't need to do extra work for loading default values.

**Comment by mcvsubbu** [comment]
There is also a case of older tables (that exist even now) when schema was optional for offline tables

**Comment by Jackie-Jiang** [comment]
> There is also a case of older tables (that exist even now) when schema was optional for offline tables

Good point. In that case, we can use the schema from the segment

**Review by Jackie-Jiang — COMMENTED** [review]
@mcvsubbu After a second thought, all other minion tasks (merge rollup, realtime to offline, segment generation) require schema to be stored in ZK, so I think we can just follow the same way, and fail the task if schema doesn't exist

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/purge/PurgeTaskExecutor.java [hunk 2]
```diff
@@ -44,16 +45,19 @@ protected SegmentConversionResult convert(PinotTaskConfig pinotTaskConfig, File
...
     SegmentPurger.RecordPurgerFactory recordPurgerFactory = MINION_CONTEXT.getRecordPurgerFactory();
     SegmentPurger.RecordPurger recordPurger =
         recordPurgerFactory != null ? recordPurgerFactory.getRecordPurger(rawTableName) : null;
     SegmentPurger.RecordModifierFactory recordModifierFactory = MINION_CONTEXT.getRecordModifierFactory();
     SegmentPurger.RecordModifier recordModifier =
         recordModifierFactory != null ? recordModifierFactory.getRecordModifier(rawTableName) : null;
 
+    TableConfig tableConfig = getTableConfig(tableNameWithType);
+    String schemaName = tableConfig.getValidationConfig().getSchemaName();
+    Schema schema = getSchema(schemaName);
```
[the remark this answers is not in the document]
Comment (sajjad-moradi) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/purge/PurgeTaskExecutor.java:57]: I'm not sure about all other places in code, but at least in this integration test, one schema is used for all three tables.
I will refactor this.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/purge/PurgeTaskExecutor.java [hunk 3]
```diff
@@ -44,16 +45,18 @@ protected SegmentConversionResult convert(PinotTaskConfig pinotTaskConfig, File
...
-    TableConfig tableConfig = getTableConfig(tableNameWithType);
     SegmentPurger.RecordPurgerFactory recordPurgerFactory = MINION_CONTEXT.getRecordPurgerFactory();
     SegmentPurger.RecordPurger recordPurger =
         recordPurgerFactory != null ? recordPurgerFactory.getRecordPurger(rawTableName) : null;
     SegmentPurger.RecordModifierFactory recordModifierFactory = MINION_CONTEXT.getRecordModifierFactory();
     SegmentPurger.RecordModifier recordModifier =
         recordModifierFactory != null ? recordModifierFactory.getRecordModifier(rawTableName) : null;
 
+    TableConfig tableConfig = getTableConfig(tableNameWithType);
+    Schema schema = getSchema(rawTableName);
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/purge/PurgeTaskExecutor.java:56]:
```suggestion
    Schema schema = getSchema(tableNameWithType);
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-core/src/main/java/org/apache/pinot/core/minion/SegmentPurger.java [hunk 4]
```diff
@@ -79,7 +82,7 @@ public File purgeSegment()
         return null;
       }
 
-      SegmentGeneratorConfig config = new SegmentGeneratorConfig(_tableConfig, segmentMetadata.getSchema());
+      SegmentGeneratorConfig config = new SegmentGeneratorConfig(_tableConfig, _schema);
```
Comment (snleee) [inline pinot-core/src/main/java/org/apache/pinot/core/minion/SegmentPurger.java:85]: If `_schema = null`, probably it's safer to fall back to `segmentMetadata.getSchema()`?

**Review by snleee — APPROVED** [review]
LGTM otherwise.

**Pull request merged**